src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java

```diff
@@ -0,0 +1,209 @@
+ * <p>
+ * The plugin provides the ability different options like choosing interactive or batch execution of sql query, setting
+ * of resulting dataset and table, enabling/disabling cache, specifying whether the query being executed is legacy or
+ * standard and retry strategy.
+ */
+@Plugin(type = Action.PLUGIN_TYPE)
+@Name(BigQueryExecute.NAME)
+@Description("Execute a Google BigQuery SQL.")
+public final class BigQueryExecute extends Action {
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryExecute.class);
+  public static final String NAME = "BigQueryExecute";
+  private static final String RECORDS_OUT = "records.out";
+
+  private Config config;
+
+  @Override
+  public void run(ActionContext context) throws Exception {
+    config.validate();
+    QueryJobConfiguration.Builder builder = QueryJobConfiguration.newBuilder(config.getSql());
+    // Run at batch priority, which won't count toward concurrent rate limit.
+    if (config.getMode().equals(QueryJobConfiguration.Priority.BATCH)) {
+      builder.setPriority(QueryJobConfiguration.Priority.BATCH);
+    } else {
+      builder.setPriority(QueryJobConfiguration.Priority.INTERACTIVE);
+    }
+
+    // Save the results of the query to a permanent table.
+    if (config.getDataset() != null  && config.getTable() != null) {
+      builder.setDestinationTable(TableId.of(config.getDataset(), config.getTable()));
+    }
+
+    // Enable or Disable the query cache to force live query evaluation.
+    if (config.shouldUseCache()) {
+      builder.setUseQueryCache(true);
+    }
+
+    // Enable legacy SQL
+    builder.setUseLegacySql(config.isLegacySQL());
+
+    QueryJobConfiguration queryConfig = builder.build();
+
+    // Location must match that of the dataset(s) referenced in the query.
+    JobId jobId = JobId.newBuilder().setRandomJob().setLocation(config.getLocation()).build();
+
+    // API request - starts the query.
+    Credentials credentials = config.getServiceAccountFilePath() == null ?
+                                null : GCPUtils.loadServiceAccountCredentials(config.getServiceAccountFilePath());
+    BigQuery bigQuery = GCPUtils.getBigQuery(config.getProject(), credentials);
+    Job queryJob = bigQuery.create(JobInfo.newBuilder(queryConfig).setJobId(jobId).build());
+
+    LOG.info("Executing SQL as job {}.", jobId.getJob());
+    LOG.debug("The BigQuery SQL is {}", config.getSql());
+
+    // Wait for the query to complete
+    queryJob.waitFor();
+
+    // Check for errors
+    if (queryJob.getStatus().getError() != null) {
+      // You can also look at queryJob.getStatus().getExecutionErrors() for all
+      // errors, not just the latest one.
+      throw new RuntimeException(queryJob.getStatus().getExecutionErrors().toString());
+    }
+
+    TableResult queryResults = queryJob.getQueryResults();
+    long rows = queryResults.getTotalRows();
+
+    context.getMetrics().gauge(context.getStageName() + RECORDS_OUT, rows);
```

Comment:
this should not be done. records.out is the metric emitted by the app indicating records flowing from one stage to another. 'records.processed' would be more accurate.

This is also doing something weird with concatenation, as there's no separator between stage name and records.out. You should use context.getStageMetrics() instead of prefixing. Also, is there a reason gauge is used instead of increment?

Reply:
gauge and increment should be the same in this case, since we are only adding the metrics once. I have fixed the metrics emit